### ISSUE TITLE
What about Kubernetes Secrets vs. environment variables support in Craft

### ISSUE THREAD

**Issue by bertoost** [craftcms/cms#5945]
Hi guys,

I want to startup a discussion about Kubernetes Secrets. Not that this is a part of the Craft CMS support, but it has something to do with environment variables, which Craft supports.

Maybe you know, maybe you don't... But Secrets in Kubernetes are used to store credentials in. Those secrets are basically put back into containers as files, mounted typically on something like `/var/run/secrets/` or whatever.
For now I have to do put them in a configuration file, since I have to read the contents of these files like `file_get_contents(getenv('MY_ENV_VAR_AS_SECRET'))`.
This way I can't use the environment variables directly in the CP env-vars-support fields like email settings or site settings etc.

For example; In Symfony you can use these kind of secrets like parameters defined as `%env(file:MY_ENV_VAR_AS_SECRET)%`. The "file" keyword is being parsed as a `file_get_contents()` would do. Maybe this is something Craft can support too. This makes working with Secrets in new environment infrastructures (like Kubernetes) and Craft even more powerful.

I hope you see the value of this and I would like to talk about this kinda feature.

### Additional info

- Craft version: 3.4.15
- PHP version: 7.4.4

**Comment by jasonmccallister** [comment]
Hi @bertoost, I really like the way Symfony handles environment variables, they have a really powerful library. Out of curiosity, is the Kubernetes cluster you are using now limited to only volume mounts for secrets? You can use Kubernetes secrets with Craft without any modifications to the way craft processes environment variables. 

There are two common methods to mounting secrets in a container. The method you mention (volume mounts) is a common method, but you are also able to use secrets as environment variables. Kubernetes has an example here: https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables.

I do agree that the Symfony environment variable library is powerful and maybe something we should consider down the line. However, the way Craft handles environment variables should not limit you from using Kubernetes secrets now.

**Comment by bertoost** [comment]
Hi @jasonmccallister 
I am aware of loading secrets through environment variables. But as far I consider security, mounting them via volumes is more secure. Therefore the benefits of using them as mounts is when you change your secrets, you don't have to reload your container(s).
See (for example) https://stackoverflow.com/questions/51365355/kubernetes-secrets-volumes-vs-environment-variables

**Comment by jasonmccallister** [comment]
I figured but just wanted to verify. I agree with your stance, a lot of people forget that a simple `phpinfo();` command can expose all the goodies. 

One thing to consider is also local development, where the environment may not match and you might have the make some hacks to easily work between environments. Although in this case, you would _should_ be running in containers locally. I have handled this in the past by using something similar (this is not complete as it should handle errors better):

```php
'securityKey' => function (): string {
    if (getenv('SECURITY_KEY')) {
        return getenv('SECURITY_KEY');
    }

    return file_get_contents('/run/secrets/security-key');
},
```

As a side note, a really awesome part of how Symfony handles environment variables is that you can extend the functionality with your own code: 
https://symfony.com/doc/current/configuration/env_var_processors.html#custom-environment-variable-processors.

**Comment by bertoost** [comment]
@jasonmccallister indeed. My local development is not the same way as my Kubernetes environment, and I literally love the way how Craft support multi-environment config files and many times I do like this;

```php
return [
    '*' => [
        'securityKey' => file_get_contents(getenv('SECURITY_KEY')),
    ],
    'dev' => [
        'securityKey' => getenv('SECURITY_KEY'),
    ],
];
```
which is fair enough the most of the time.. But...
when you decide to use the CP feature of using environment variables, you're not able to use Secrets from Kubernetes environment inside the CP use of environment variables.

For example;
When I configure the mail adapter, I configure it for production, using environment variables. Unfortunate I have to parse them manually like this example for my own Mailjet adapter

```php
// Prod environment settings
    'prod' => [
        'components' => [
            'mailer' => static function () {
                $settings = craft\helpers\App::mailSettings();

                if ($settings->transportType === \bertoost\mailjet\adapters\MailjetAdapter::class) {
                    $settings->transportSettings['apiKey'] = file_get_contents(Craft::parseEnv($settings->transportSettings['apiKey']));
                    $settings->transportSettings['apiSecret'] = file_get_contents(Craft::parseEnv($settings->transportSettings['apiSecret']));
                }

                $config = craft\helpers\App::mailerConfig($settings);

                return Craft::createObject($config);
            },
        ],
    ],
```

If Craft could support, somehow, to render environment variables as files.. above code-snippet can be removed for such setups.
For example; when Craft checks, if the environment variable is a physically existing file before using the value directly, it would save a lot of extra code you have to write yourself

**Comment by bertoost** [comment]
Oh and yes.. I love Symfony's extendability, it's very flexible.. but from a CMS point of view; Craft is by far my most favorite all times 😄  (until now hehe)
